File: 21_Merge_Two_Sorted_Lists/solution.cpp
**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        /*
            Inputs: head of linked list list1 and list2
            Output: head of the merged linked list

            Have a dummy head for ease of returning the start of the linked list
            - compare the values of the two nodes from each list
            while node 1 and node 2:
            - if node 1 <= node 2:
                - add node 1 to the dummy head
                - progress node 1
            - else
                - add node 2 to the dummy head
                - progress node 2
            after exiting the loop, check for remaining nodes either in list 1 or list 2 and append it to dummy head

            ListNode *dummy = new ListNode(0);
            ListNode *curr = dummy;
            while (list1 && list2) {
                if (list1->val <= list2->val) {
                    curr->next = list1;
                    curr = curr->next;
                    list1 = list1->next;
                }else{
                    curr->next = list2;
                    curr = curr->next;
                    list2 = list2->next;
                }
            }
            if(list1){
                curr->next = list1;
            }else if(list2){
                curr->next = list2;
            }
            return dummy->next;

                

        */
        ListNode temp = ListNode(0);
        ListNode *dummy = &temp;
        ListNode *curr = dummy;
        while (list1 && list2) {
            if (list1->val <= list2->val) {
                curr->next = list1;
                curr = curr->next;
                list1 = list1->next;
            }else{
                curr->next = list2;
                curr = curr->next;
                list2 = list2->next;
            }
        }
        if(list1){
            curr->next = list1;
        }else if(list2){
            curr->next = list2;
        }
        return dummy->next;
    }
};

File: 21_Merge_Two_Sorted_Lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        /*
            Input: heads of two sorted linked lists
            Output: head of joined sorted linked list

            Strategy
            - initialize a dummy head, have another reference pointing to it
            - while current node of both linked list are not null
                - if value of node from list 1 <= value of node from list 2
                    - join node from list 1 to the dummy head
                    - advance dummy head and node from list 1
                - else
                    - join node from list 2 to the dummy head
                    - advance dummy head and node from list 2
            - When the while loop end
                - if node from list 1 is not null
                    - join it to dummy head 
                - else if node from list 2 is not null
                    - join it to dummy head
                return copy reference.next
        */
        ListNode dummy = new ListNode();
        ListNode res = dummy;

        while(list1 != null && list2 != null) {
            if(list1.val <= list2.val) {
                dummy.next = list1;
                list1 = list1.next;
            }else{
                dummy.next = list2;
                list2 = list2.next;
            }
            dummy = dummy.next;
        }

        if(list1 != null) {
            dummy.next = list1;
        }
        if(list2 != null) {
            dummy.next = list2;
        }
        return res.next;
    }
}
